File: src/Visitor.php
<?php

namespace Ann;

use \Ann\Network;
use \Ann\Neuron;
use \Ann\Branch;
use \Ann\OutputFunction;
use \Ann\Peripheral;
use \Ann\Bias;
use \Ann\Dendrite;
use \Ann\Synapse;

interface Visitor
{
    public function visitNetwork(Network $network, array $inputs, array $outputs);
    public function visitNeuron(Neuron $neuron, Branch $branch, OutputFunction $function);
    public function visitPeripheral(Peripheral $peripheral);
    public function visitBias(Bias $bias);
    public function visitDendrite(Dendrite $dendrite, array $synapses);
    public function visitSynapse(Synapse $synapse, Neuron $neuron, $weight);
}

File: ext/ann/visitor.c

#ifdef HAVE_CONFIG_H
#include "../ext_config.h"
#endif

#include <php.h>
#include "../php_ext.h"
#include "../ext.h"

#include <Zend/zend_exceptions.h>

#include "kernel/main.h"


ZEPHIR_INIT_CLASS(Ann_Visitor) {

	ZEPHIR_REGISTER_INTERFACE(Ann, Visitor, ann, visitor, ann_visitor_method_entry);

	return SUCCESS;

}

ZEPHIR_DOC_METHOD(Ann_Visitor, visitNetwork);

ZEPHIR_DOC_METHOD(Ann_Visitor, visitNeuron);

ZEPHIR_DOC_METHOD(Ann_Visitor, visitPeripheral);

ZEPHIR_DOC_METHOD(Ann_Visitor, visitBias);

ZEPHIR_DOC_METHOD(Ann_Visitor, visitDendrite);

ZEPHIR_DOC_METHOD(Ann_Visitor, visitSynapse);

